oap-server/server-storage-plugin/storage-banyandb-plugin/src/main/java/org/apache/skywalking/banyandb/client/impl/BanyanDBGrpcClient.java

```diff
@@ -0,0 +1,173 @@
+                .setMetadata(Database.Metadata.newBuilder().setGroup(traceFetchRequest.getGroup()).setName(traceFetchRequest.getName()).build())
+                .addFields(Query.PairQuery.newBuilder()
+                        .setOp(Query.PairQuery.BinaryOp.BINARY_OP_EQ)
+                        .setCondition(Query.TypedPair.newBuilder().setStrPair(Query.StrPair.newBuilder().setKey("trace_id").addValues(traceFetchRequest.getTraceId()).build()))
+                        .build())
+                .setProjection(Query.Projection.newBuilder()
+                        .addAllKeyNames(traceFetchRequest.getProjections())
+                        .build());
+        return convertToResponse(this.stub.query(queryBuilder.build()));
+    }
+
+    static BanyanDBQueryResponse convertToResponse(Query.QueryResponse response) {
+        BanyanDBQueryResponse r = new BanyanDBQueryResponse();
+        r.setTotal(response.getEntitiesCount());
+        for (final Query.Entity entity : response.getEntitiesList()) {
+            BanyanDBEntity.BanyanDBEntityBuilder entityBuilder = BanyanDBEntity.builder();
+            for (final Query.TypedPair typedPair : entity.getFieldsList()) {
+                if (typedPair.hasIntPair()) {
+                    final Query.IntPair pair = typedPair.getIntPair();
+                    entityBuilder.field(pair.getKey(), pair.getValuesList().get(0));
+                } else if (typedPair.hasStrPair()) {
+                    final Query.StrPair pair = typedPair.getStrPair();
+                    entityBuilder.field(pair.getKey(), pair.getValuesList().get(0));
+                }
+            }
+            entityBuilder.entityId(entity.getEntityId());
+            entityBuilder.binaryData(entity.getDataBinary().toByteArray());
+            entityBuilder.timestampSeconds(entity.getTimestamp().getSeconds());
+            entityBuilder.timestampNanoSeconds(entity.getTimestamp().getNanos());
+            r.getEntities().add(entityBuilder.build());
+        }
+
+        return r;
+    }
+
+    @Override
+    public void writeEntity(List<TraceWriteRequest> data) {
+        StreamObserver<org.apache.skywalking.banyandb.Write.WriteRequest> observer = this.asyncStub.withDeadlineAfter(30, TimeUnit.SECONDS)
+                .write(new StreamObserver<Write.WriteResponse>() {
+                    @Override
+                    public void onNext(Write.WriteResponse writeResponse) {
+                    }
+
+                    @Override
+                    public void onError(Throwable throwable) {
+                        log.error("fail to send request", throwable);
+                    }
+
+                    @Override
+                    public void onCompleted() {
+                    }
+                });
+
+        for (final TraceWriteRequest entity : data) {
+            try {
+                Write.EntityValue entityValue = Write.EntityValue.newBuilder()
+                        .addAllFields(buildWriteFields(entity.getFields()))
```

Comment:
Don't use like this. `get whole list`, then `iterating`, and `instance of` for every element.

Reply:
Done. Use typed API